components/clarity-lsp/src/vscode_bridge.rs

```diff
@@ -114,54 +83,83 @@ impl LspVscodeBridge {
+                };
+                let uri = &params.text_document.uri;
+
+                if let Some(contract_location) = get_contract_location(uri) {
+                    LspNotification::ContractChanged(
+                        contract_location,
+                        params.content_changes[0].text.to_string(),
+                    )
+                } else {
+                    return Promise::resolve(&JsValue::FALSE);
+                }
+            }
+
+            DidCloseTextDocument::METHOD => {
+                let params: DidCloseTextDocumentParams = match decode_from_js(js_params) {
+                    Ok(params) => params,
+                    Err(err) => return Promise::reject(&JsValue::from(format!("error: {}", err))),
                 };
+                let uri = &params.text_document.uri;
 
-                let editor_state_lock = self.editor_state_lock.clone();
-                let send_diagnostic = self.client_diagnostic_tx.clone();
-
-                return future_to_promise(async move {
-                    let mut result = match editor_state_lock.try_write() {
-                        Ok(mut editor_state) => {
-                            process_notification(command, &mut editor_state, Some(&file_accessor))
-                                .await
-                        }
-                        Err(_) => return Err(JsValue::from("unable to lock editor_state")),
-                    };
-
-                    let mut aggregated_diagnostics = vec![];
-
-                    if let Ok(ref mut response) = result {
-                        aggregated_diagnostics.append(&mut response.aggregated_diagnostics);
-                    }
-
-                    for (location, mut diags) in aggregated_diagnostics.into_iter() {
-                        if let Ok(uri) = Url::parse(&location.to_string()) {
-                            let value = PublishDiagnosticsParams {
-                                uri,
-                                diagnostics: clarity_diagnostics_to_lsp_type(&mut diags),
-                                version: None,
-                            };
-
-                            send_diagnostic.call1(&JsValue::NULL, &encode_to_js(&value)?)?;
-                        }
-                    }
-
-                    Ok(JsValue::TRUE)
-                });
+                if let Some(contract_location) = get_contract_location(uri) {
+                    LspNotification::ContractClosed(contract_location)
+                } else {
+                    return Promise::resolve(&JsValue::FALSE);
+                }
             }
+
             _ => {
                 #[cfg(debug_assertions)]
                 log!("unexpected notification ({})", method);
+                return Promise::resolve(&JsValue::FALSE);
             }
-        }
+        };
 
-        return Promise::resolve(&JsValue::NULL);
+        let mut editor_state = self.editor_state.clone();
```

Comment:
I tried to avoid the `mut` here because it seems to be unnecessary to do it on the  RwLock.
But `process_notification` needs to accept `&mut EditorState` since the `native_bridge` directly passes a mutable reference of the editor_state.

Maybe there's a way around in the way EditorStateInput is implemented but could find it

Reply:
I got confused with the naming here, see suggestion above.